lib/main.js

```diff
@@ -493,28 +513,65 @@ const Hydrogen = {
       cells,
       ({ start, end }: { start: atom$Point, end: atom$Point }) => {
         const code = codeManager.getTextInRange(editor, start, end);
-        const endRow = codeManager.escapeBlankRows(editor, start.row, end.row);
-        this._createResultBubble(editor, kernel, code, endRow);
+        const row = codeManager.escapeBlankRows(
+          editor,
+          start.row,
+          end.row == editor.getLastBufferRow() ? end.row : end.row - 1
+        );
+        const cellType = codeManager.getMetadataForRow(editor, start);
+        this._createResultBubble(editor, kernel, { code, row, cellType });
       }
     );
   },
 
   runAllAbove() {
-    const editor = store.editor; // to make flow happy
-    if (!editor) return;
+    const { editor, kernel, grammar, filePath } = store;
+    if (!editor || !grammar || !filePath) return;
     if (isMultilanguageGrammar(editor.getGrammar())) {
       atom.notifications.addError(
         '"Run All Above" is not supported for this file type!'
       );
       return;
     }
 
-    const cursor = editor.getLastCursor();
-    const row = codeManager.escapeBlankRows(editor, 0, cursor.getBufferRow());
-    const code = codeManager.getRows(editor, 0, row);
+    if (editor && kernel) {
+      this._runAllAbove(editor, kernel);
+      return;
+    }
+
+    kernelManager.startKernelFor(
+      grammar,
+      editor,
+      filePath,
+      (kernel: ZMQKernel) => {
+        this._runAllAbove(editor, kernel);
+      }
+    );
+  },
 
-    if (code) {
-      this.createResultBubble(editor, code, row);
+  _runAllAbove(editor: atom$TextEditor, kernel: Kernel) {
+    const cursor = editor.getCursorBufferPosition();
+    cursor.column = editor.getBuffer().lineLengthForRow(cursor.row);
+    const breakpoints = codeManager.getBreakpoints(editor);
+    breakpoints.push(cursor);
+    const cells = codeManager.getCells(editor, breakpoints);
+    for (let i = 0; i < cells.length; i++) {
```

Comment:
Might be nitpicky, but you could use `for...of`:

```js
for (const cell of cells) {
      const { start, end } = cell;
      log("cell", cell);
...
```

We could probably drop the `log` there too if just for debugging.

Reply:
This made two functions much nicer and cleaner. Thank you